File: Solutions/39. Combination Sum/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        vector<int> tmp;
        sort(nums.begin(),nums.end());
        dfs(res,tmp,nums,target,0);
        return res;
    }
    void dfs(vector<vector<int>>& res,vector<int>& tmp,vector<int>& nums,int target,int st){
        if(target<0){
            return;
        }
        if(target==0){
            res.push_back(tmp);
            return;
        }
        for(int i=st;i<nums.size();++i){
            tmp.push_back(nums[i]);
            dfs(res,tmp,nums,target-nums[i],i);
            tmp.pop_back();
        }
    }
};

File: Solutions/39. Combination Sum/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res=new ArrayList<>();
        List<Integer> path=new ArrayList<>();
        foo(candidates,target,path,res,0);
        return res;
    }
    private void foo(int[] c,int t,List<Integer> path,List<List<Integer>> res,int begin){
        if(t<=0){
            if(t==0){
                res.add(new ArrayList<>(path));
                return;
            }
            return;
        }
        for(int i=begin;i<c.length;++i){
            path.add(c[i]);
            foo(c,t-c[i],path,res,i);
            path.remove(path.size()-1);
        }
    }
}